[CALCITE-7010] The well-known count bug

suibianwanwank opened the pull request:
This PR combines an existing implementation of Calcite with the paper [Improving Unnesting of Complex Queries](https://15799.courses.cs.cmu.edu/spring2025/papers/11-unnesting/neumann-btw2025.pdf) to fix the classic count-bug when decorrelated subqueries. For more information, see [CALCITE-7010].
Pseudo-code given in the paper:
```
1 fun unnest(groupby, info, accessing):
2 static = groupby.groups is empty or groups.groupingsets contains ∅
3 unnest(groupby.input, info, accessing)
4 rewriteColumns(groupby, info)
5 for c in info.outerRefs:
6 add info.repr[c] to groupby.groups
7 if static:
8 replace groupby with info.D groupy, joining on mapped info.outerRefs
```
This fix affects a lot of existing test plans, and I haven't yet fully fixed it and evaluated its impact. However, based on the sub-query.iq tests, the results are very promising.
 I'll set this PR to draft until all tests are fixed. Any insight on this would be appreciated!

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -759,6 +758,68 @@ protected RexNode removeCorrelationExpr(
     }
 
     RelNode newRel = relBuilder.build();
+    // Special case where the group by is static (i.e., a select clause with aggregation functions
```
why "static"?
there really is only a single group containing all elements.
alternatively, this is the same as grouping on 'true' (or any constant).

suibianwanwank replied:
Actually, this isn't exactly the same:
```
select sum(i0) from  (select i0 from test where i1 = 1) a

sum |
----+
null|
(1 rows)
```

```
select sum(i0) from  (select i0 from test where i1 = 1) a group by true

sum|
---+
(0 rows)
```

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -759,6 +758,68 @@ protected RexNode removeCorrelationExpr(
+      final Frame outFrame = outerFramePair.right;
+      RexBuilder rexBuilder = relBuilder.getRexBuilder();
+
+      int groupSize = (int) corDefOutputs.keySet().stream()
```
is "keySize" a better name?
or groupKeySize?

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -759,6 +758,68 @@ protected RexNode removeCorrelationExpr(
+          .count();
+      List<RelDataTypeField> newRelFields = newRel.getRowType().getFieldList();
+      ImmutableBitSet.Builder corFieldBuilder = ImmutableBitSet.builder();
+      final Map<Integer, RexNode> specificRexNodeMap = new HashMap<>();
```
can you describe the keys and values of this map?
seems to map an index to a field access expression.

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -759,6 +758,68 @@ protected RexNode removeCorrelationExpr(
+      final Map<Integer, RexNode> specificRexNodeMap = new HashMap<>();
+      final List<RexNode> conditions = new ArrayList<>();
+      for (Map.Entry<CorDef, Integer> corDefOutput : corDefOutputs.entrySet()) {
+        if (corDefOutput.getKey().corr.equals(outerFramePair.left)) {
```
I think the code is easier to read if you store coreDefOutput.getKey() and getValue() in separate variables with descriptive names.

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -759,6 +758,68 @@ protected RexNode removeCorrelationExpr(
+
+      for (int i1 = 0; i1 < rel.getAggCallList().size(); i1++) {
+        AggregateCall aggCall = rel.getAggCallList().get(i1);
+        if (aggCall.getAggregation() instanceof SqlCountAggFunction) {
```
does the argument to count matter? i.e., `*`?
the paper seemed to imply it does

mihaibudiu commented on core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java:
```diff
@@ -1781,7 +1781,7 @@ private void checkSemiOrAntiJoinProjectTranspose(JoinRelType type) {
             CoreRules.PROJECT_MERGE)
         .withRule(CoreRules.PROJECT_TO_SEMI_JOIN)
-        .check();
+        .checkUnchanged();
```
why are these unchanged now?
This suggests that decorrelation fails in this case.

suibianwanwank replied:
It appears the SemiJoinRule is being tested here. After the new decorrelation plan, this rule was not successfully applied. I'll try to research this more thoroughly.

mihaibudiu commented on core/src/test/java/org/apache/calcite/test/SqlHintsConverterTest.java:
```diff
@@ -519,48 +519,6 @@ public final Fixture sql(String sql) {
         .check();
   }
 
-  @Test void testHintsPropagationInVolcanoPlannerRules2() {
```
I am not sure it's ok to delete such tests without a very good reason

suibianwanwank replied:
I'm not sure why we're using a correlated query here to test hint passing. maybe I should change the original SQL to a more explicit join.

mihaibudiu commented on core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:
```diff
@@ -1003,16 +1003,22 @@ and e1.sal > (select avg(e2.sal) from emp e2
-          LogicalProject(DEPTNO=[$2], EXPR$0=[$3])
-            LogicalAggregate(group=[{0, 1, 2}], groups=[[{0, 1, 2}, {0, 2}, {1, 2}, {2}]], EXPR$0=[AVG($3)])
-              LogicalProject(COMM=[$6], MGR=[$3], DEPTNO=[$7], SAL=[$5])
+        LogicalProject(DEPTNO0=[$0], $f1=[$2])
```
were the previous plans incorrect?

suibianwanwank replied:
From the plan, the original plan was also correct. The root cause of the `count bug` is that during Aggregate processing, the `corVar` field gets added to GROUP BY for deduplication and is used in subsequent JOINs. For `sum(a)`, when the input has 0 rows, it returns null, but for `sum(a) GROUP BY b`, it returns 0 rows. This is what the PR fixes. However, if a parent RelNode returns identical results for null columns and empty rows, we actually don't need extra processing—though currently there's no good way to determine this. Perhaps we could handle this case in the rule.

mihaibudiu reviewed (commented):
I will have to read the paper and the changes more carefully to make truly informed comments.
But if this is a bug, it has to be fixed.
The compiler should either refuse to convert a query or produce a correct result, but never produce an incorrect result from a rewrite.

caicancai commented:
So cool, I will read the paper and review it.

suibianwanwank commented:
> Can we revive this PR?

I think it's ready~

suibianwanwank commented on core/src/test/java/org/apache/calcite/rel/rel2sql/RelToSqlConverterTest.java:
```diff
@@ -7378,20 +7378,6 @@ private void checkLiteral2(String expression, String expected) {
         callsUnparseCallOnSqlSelect[0], is(true));
   }
 
-  @Test void testCorrelate() {
```
Keeping this test seems pointless, as it attempts to make decorrelate fail using value, but this PR already supports decorrelation for value.

suibianwanwank commented on core/src/test/resources/org/apache/calcite/test/RelOptRulesTest.xml:
```diff
@@ -18919,13 +18922,16 @@ LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$
       <![CDATA[
 LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8])
-  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8], m=[CAST($9):INTEGER], c=[CAST($10):BIGINT], d=[CAST($11):BIGINT], trueLiteral=[CAST($12):BOOLEAN], NAME=[CAST($13):VARCHAR(10)])
+  LogicalProject(EMPNO=[$0], ENAME=[$1], JOB=[$2], MGR=[$3], HIREDATE=[$4], SAL=[$5], COMM=[$6], DEPTNO=[$7], SLACKER=[$8], m=[$9], c=[CAST($10):BIGINT], d=[CAST($11):BIGINT], trueLiteral=[$12], JOB0=[CAST($13):VARCHAR(10)])
```
Fixed, I kept the aliases.

suibianwanwank commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -759,6 +758,68 @@ protected RexNode removeCorrelationExpr(
     }
 
     RelNode newRel = relBuilder.build();
+    // Special case where the group by is static (i.e., a select clause with aggregation functions
```
Done. Thanks!

suibianwanwank commented:
Thanks for the review! I've updated some comments and added statements for clarity. PTAL @mihaibudiu @iwanttobepowerful .

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -760,6 +766,177 @@ protected RexNode removeCorrelationExpr(
+    // Rewrite Aggregate as:
+    //   SELECT d.deptno FROM dept d
+    //     JOIN (
+    //          SELECT true AS cs, d2.deptno
```
had to remove this "d2" to make this program compile

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -760,6 +766,177 @@ protected RexNode removeCorrelationExpr(
+      }
+
+      ImmutableBitSet groupSet = corFieldBuilder.build();
+      final RelNode join = relBuilder.push(outFrame.r)
```
you can annotate the code here with the line from the above plan that is being constructed

mihaibudiu commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -760,6 +766,177 @@ protected RexNode removeCorrelationExpr(
+    //          WHERE cnt = 0
+    //     ) AS d0 ON d.deptno = d0.deptno
+    // Corresponding plan:
+    //    LogicalProject(DEPTNO=[$0])
```
I suggest numbering these lines - see below.

mihaibudiu reviewed (commented):
I think I am starting to understand what's going on; made some suggestions on making the code easier to read. I will review again.

suibianwanwank commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -760,6 +766,177 @@ protected RexNode removeCorrelationExpr(
+    //          SELECT true AS cs, d2.deptno
+    //          FROM (
+    //              SELECT d2.deptno,
+    //                     CASE WHEN cnt0 IS NOT NULL THEN cnt0 ELSE 0 END AS cnt
```
Yes, to be more precise, for any join key in subsequent join, `count` needs to return 0, and other aggregate functions must return NULL. (NULL is also not the same as an empty set—if there's a filter like `is not null`, so the count-bug isn't limited to just `count`.)

suibianwanwank commented:
@mihaibudiu I've addressed all the comments. I modified it into an additional method because the excessive number of lines in the method caused the CI to fail (though most of them were comments).

suibianwanwank commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -760,11 +766,198 @@ protected RexNode removeCorrelationExpr(
+   *          LogicalTableScan(table=[[scott, DEPT]])
+   *        LogicalProject(cs=[true])
+   *          LogicalFilter(condition=[=(0, $0)])
+   *            LogicalAggregate(group=[{}], EXPR$0=[COUNT()])
```
Done.

suibianwanwank commented on core/src/test/resources/sql/scalar.iq:
```diff
@@ -301,6 +301,7 @@ from (values (1), (3)) t1(id);
 
 !ok
 
+!if (false) {
```
@mihaibudiu I merged main and squashed the commits. Unfortunately, this test failed.

Before this PR, Value couldn't be decorrelated, but I believe this issue is unrelated to the current PR because replacing Value with a table would still trigger the same error. Additionally, I'm confident this issue can be fixed by [CALCITE-7045], and I'll start working on it after this PR is merged.

suibianwanwank commented:
If there are no other comments, I will merge this PR today.

suibianwanwank commented on core/src/main/java/org/apache/calcite/sql2rel/RelDecorrelator.java:
```diff
@@ -782,11 +788,198 @@ protected RexNode removeCorrelationExpr(
+   * [09]             LogicalAggregate(group=[{0}])
+   * [10]               LogicalProject(DEPTNO=[$0])
+   * [11]                 LogicalTableScan(table=[[scott, DEPT]])
+   * [12]             LogicalAggregate(group=[{0}], EXPR$0=[COUNT()])
```
No, this does not include any optimizations.  I guess the following comment explains the issue you're wondering about:
```
    // Build [09] LogicalAggregate(group=[{0}]) to obtain the distinct set of
    // corVar from outFrame.
```
The Aggregate here is for distinct purposes.

suibianwanwank replied:
My oversight, thanks. I've adjusted the SQL stmt.
